Monetization API integration

mxr576 opened the pull request:
## [Ready for review]

Install it as patch with composer-patches if you need it!

```json
    "require": {
        "cweagans/composer-patches": "^1.6.5",
    },
    "config": {
        "preferred-install": {
            "*": "dist",
            "apigee/apigee-client-php": "source"
        },
    },
    "patches": {
            "apigee/apigee-client-php": {
                "[WIP] Monetization API integration - phase 1": "https://github.com/apigee/apigee-client-php/pull/18.diff"
            }
    },
````

mxr576 committed:
Monetization API integration - phase 1.

mxr576 committed:
Monetization API integration tests - phase 1.

mxr576 commented:
@Jaesin Thanks for the hands up, I'll fix this next week to return either null or BalanceInterface. Writing tests for the existing implementations in this PR is still in progress and until that time these type of issues can still occur.

mxr576 committed:
Fixing CS and return type issues.

mxr576 committed:
Monetization API integration tests - phase 1, part 2.

mxr576 committed:
Company and developer controller tests.

mxr576 committed:
SupportedCurrencies and PaymentProviders tests plus other improvements
on existing tests.

mxr576 committed:
Fixed pagination.

mxr576 committed:
Setters become internal in legal entities and additional adjustments on
the related tests.

mxr576 committed:
Major changes around entity serialization.

Refactored entity serialization/normalization/denormalization logic,
fixed some issues around rate plans (there are still some), more tests!

mxr576 committed:
Fixed issue with standard rate plan serialization.

mxr576 committed:
Terms and Conditions API support with tests.

mxr576 committed:
Added support to developer accepted rate plans.

TODO Testing - after we got some answers from engineering.

mxr576 committed:
More rate plan tests for different rate plan types.

mxr576 committed:
Fixing the name of some test classes.

mxr576 committed:
Rate plan revision tests.

mxr576 committed:
Test rate plan listing and new revision creation.

mxr576 committed:
Fixing incorrect class name.

mxr576 committed:
Separated "active" and "accepted" rate plan plan related implementations
. Extended test coverage with tests for the related classes.

mxr576 committed:
Changed assertion on some properties.

mxr576 committed:
Extended test files to improve test coverage.

mxr576 committed:
Fixed failing tests around accepted rate plans.

mxr576 commented:
@Jaesin Where do you see it exactly in the response? I think it is a property of a rate plan revision so you should get access to its value with if the accepted rate plan is a rate plan revision.

```php
//....
if ($acceptedRatePlan->getRatePlan() instanceOf RatePlanRevisionInterface) {
   echo $acceptedRatePlan->getRatePlan()->isKeepOriginalStartDate() ? 'true' : 'false';
}
```

https://github.com/mxr576/apigee-client-php/blob/mint/tests/Api/Monetization/EntitySerializer/RatePlanSerializerValidator.php#L67

https://docs.apigee.com/api-platform/monetization/create-future-rate-plans#setupfrpapi

mxr576 committed:
Prepaid balances tests and extended code coverage.

mxr576 committed:
Added missing file.

mxr576 committed:
Fix failing tests on PHP 7.2.

mxr576 committed:
Fixed content of the test file.

mxr576 committed:
Renamed a few test classes.

mxr576 committed:
Updated Monetization API tests to use the new test API- and HTTP
clients.

mxr576 committed:
Updated Monetization API tests to use the new static
TestClientFactory::getClient().

mxr576 committed:
Explode entity CRUD operation trait and interface to smaller components
per operation.

mxr576 committed:
Updated Monetization API related controllers and tests to leverage the
exploded entity CRUD operation traits and interfaces.

Related commit: a318d37c7523.

Jaesin commented:
@mxr576 I see `keepOriginalStartDate` in `weather_forcast-vd4k::weather_forcast-vd4k_weather_per_transaction_1870225007` which is a revision.

Jaesin commented on src/Api/Monetization/Entity/Property/StartDatePropertyAwareTrait.php:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * @inheritdoc
+     */
+    public function getStartDate(): \DateTimeImmutable
```
One issue I've had is when `\Drupal\apigee_edge\Entity\FieldableEdgeEntityBaseTrait::getOriginalFieldData` gets called on a new entity, there won't be any field values set yet.  This causes a `TypeError`. `::set` calls `::get` which calls `::getOriginalFieldData`.

mxr576 replied:
@Jaesin This is a Drupal 8 integration related issue probably, so if you need help about it please report it in the Drupal 8 module repo or in a private channel. We also had our fight with Drupal 8's typed data system when we worked on the Management API integration.

Jaesin commented on src/Api/Monetization/Normalizer/ApiPackageNormalizer.php:
```diff
@@ -0,0 +1,70 @@
+        // Do not send redundant API product information to Apigee Edge, the
+        // id of a referenced API product is enough.
+        foreach ($normalized['product'] as $id => $data) {
+            $normalized['product'][$id] = (object) ['id' => $data->id];
```
I'm getting an  error because `data->id` is not set.  I think name is used for this object.

<img width="358" alt="image" src="https://user-images.githubusercontent.com/159693/47177700-00019600-d2ce-11e8-8b6b-9f89e74db3b6.png">

mxr576 replied:
What I can see is a Drupal (Mint?) API product entity (?), why API package normalizer would be called on that if everything configured properly? The line you commented gets called when a new API package gets created with a selected list of API products.

@Jaesin do you still have this issue?

Jaesin replied:
Hi @mxr576,
Sorry for not giving better context. I may have been using a management product as a package product. Would I have to create a management product then save it and fetch the monetizaton product in order to construct an API package?

mxr576 replied:
Yes, you should do that, if you check the annotation on the trait used by the ApiPackage entity you can see that: https://github.com/apigee/apigee-client-php/pull/18/files#diff-81c72e99743d59cd652977609a66ca0cR28.
The MINT API returns nested MINT API products not MGMT API products when you are retrieving an API package. This is the reason why you should use them when you constructing an API package.

Jaesin commented on src/Api/Monetization/Denormalizer/DeveloperRatePlanDenormalizer.php:
```diff
@@ -0,0 +1,69 @@
+        }
+
+        if (parent::supportsDenormalization($data, $type, $format)) {
+            return RatePlanInterface::TYPE_DEVELOPER == $data->type && !$data->developer->isCompany;
```
I had an issue when getting all plans of a package where a developer plan was returned without the developer object causing an error here.
```
    $rate_plan_controller  = new RatePlanController('multiple_products-c4xp', $this->sdk_connector->getOrganization(), $this->sdk_connector->getClient());
    $entities = $rate_plan_controller->getEntities(FALSE);
```
Result:
```
Undefined property: stdClass::$developer
 .../vendor/apigee/apigee-client-php/src/Api/Monetization/Denormalizer/DeveloperRatePlanDenormalizer.php:64
```

mxr576 replied:
Thanks, the sample code was really useful to debug this issue. I can see the problem now but I do not understand why the developer is not being returned for a _developer_ rate plan. I'd probably need some help from engineering to answer this question.
What I can see the problem rate plan (multiple_products-c4xp_minimum_rateplan6) has expired but that should not justify this problem. I also did not find a developer called "John Doe" in the org, which could be one explanation...

mxr576 committed:
Added missing method to the interface.

mxr576 committed:
Changed method description.

mxr576 committed:
Introducing new random generator wrappers for testing.

mxr576 committed:
Introduce entity serializer and property validators.

mxr576 committed:
Refactored basic and Management API tests.

mxr576 committed:
Refactored Monetization API tests.

mxr576 committed:
Introduced random url generation feature to random generator.

mxr576 committed:
Finished implementation of the accepted terms and conditions related
classes.

mxr576 committed:
Added tests to getEntities() methods in TermsAndConditionsController.

mxr576 committed:
Leverage base entity endpoint.

mxr576 committed:
Variable name- and text changes.

mxr576 committed:
Do not send redundant query parameters, fallback to API endpoint's
default (by default).

mxr576 committed:
Nested org reference object does not contain the address of the org.

mxr576 committed:
Introduced new optional $client parameter method for Monetization API
tests.

mxr576 committed:
Fixed implementation of the active rate plan controller, added extra
tests.

mxr576 committed:
Improved implementation of developer/company accepted/declied terms and
conditions API's implementation. Added tests.

cnovak commented:
@mxr576 this PR is huge, can you break this into smaller PRs so that we can get a stable feature branch to work off of?  You can add more PRs over time as new resources in the Apigee API are completed.

mxr576 commented:
@cnovak I think we are close to merge this PR to the apigee:2.x-mint branch. I am continuously backporting the MGMT API integration related improvements from this PR to the apigee:2.x branch and with that reducing the size of this PR.

Since only ~2 user stories are remaining from this sprint I would wait until those gets resolved. (Hopefully in this week.)

mxr576 committed:
Extra sanity check that ensures nullable properties are still set when
a new entity gets serialized.

(When a new entity object constructed by a developer gets serialized.)

cnovak commented:
Hello @mxr576 , I am getting the following error:
```
Fatal error: Interface 'Apigee\Edge\Controller\EntityUpdateControllerOperationInterface' not found in /app/vendor/apigee/apigee-client-php/src/Controller/EntityCrudOperationsControllerInterface.php on line 24
```
The `EntityUpdateControllerOperationInterface` interface file is named differently than the actual interface name, words "Controller" and "Operation" are reversed in filename: 

`src/Controller/EntityUpdateOperationControllerInterface.php`

mxr576 committed:
Decouple time zone fixing feature from normalizers and denormalizers
to a shared trait.

mxr576 committed:
Fixing names of some interfaces.

mxr576 commented:
@cnovak thanks for noticing this, weird that it has not been discovered by tests.

https://github.com/apigee/apigee-client-php/pull/18/commits/8bb15435831eedb637f9d11120c3d144734bfa18 this commit should fix it.

mxr576 committed:
Apply the same type safe trick on Monetization related data objects.

mxr576 committed:
Fixing incorrect annotation on a MINT property trait.

mxr576 committed:
Added missing tests and fixed some edge cases in the
PropertyAccessorDecorator.

mxr576 commented:
@cnovak PR with Management API improvements back-ported from the Monetization API integration is here: https://github.com/apigee/apigee-client-php/pull/25

If it gets merged to 2.x branch and I rebase my mint branch then the size of this PR gets much smaller and it will only contain MINT related changes: https://github.com/mxr576/apigee-client-php/compare/mint-backport...mxr576:mint-rebase?expand=1

Jaesin commented:
I'm wondering if there are any improvements blocking this PR from becoming a feature branch on `apigee/apigee-client-php` and what still need to be done to get this into `apigee/apigee-client-php:2.x`

mxr576 commented:
@Jaesin this repo should not contain feature branches, only release branches. Feature branches should live in the forks. I created is a placeholder destination branch for this PR a few months ago: https://github.com/apigee/apigee-client-php/tree/2.x-mint.
Although, as I wrote to @cnovak today maybe we can merge this directly to the 2.x branch thanks for the 90% test coverage that I managed to achieve with offline tests. However, I still would like to ask the team to reconsider moving MINT integration to an individual PHP lib because of the same reasons why the https://github.com/apigee/apigee-m10n-drupal is a separate module. This PR clearly demonstrates the size of the MINT integration and the complexity that it adds to the MGMT API integration that already lives in the 2.x.

mxr576 commented:
Rebased the PR, keeping the old branch for a while here: https://github.com/mxr576/apigee-client-php/tree/mint-old

mxr576 committed:
Merge remote-tracking branch 'upstream/2.x' into mint

mxr576 committed:
Make idProperty() static.

cnovak commented:
We will sometimes need feature branches to collaborate on new features @mxr576 , and to not have hundreds of file changes in a PR again.  Do you have an ETA on when this will be merged into 2.x?

mxr576 commented:
@cnovak Now the day has come! \o/ Sorry, it was a little bit too late on Friday when we had finish with our call and in general I do not prefer Friday night deployments (nothing good has ever come out of them :) ).

cc @Jaesin ^
